daffodil-tdml-processor/src/main/scala/org/apache/daffodil/tdml/processor/DaffodilTDMLDFDLProcessor.scala

```diff
@@ -311,15 +323,62 @@ class DaffodilTDMLDFDLProcessor private (private var dp: DataProcessor) extends
     xri.parse(sis)
     val actual = dp.parse(dis, outputter)
 
-
     if (!actual.isError && !errorHandler.isError) {
-      verifySameParseOutput(outputter, outputStream)
+      verifySameParseOutput(outputter, saxOutputStream)
     }
-    verifySameDiagnostics(actual, errorHandler)
+    val dpParseDiag = actual.getDiagnostics.map(_.getMessage())
+    val saxParseDiag = errorHandler.getDiagnostics.map(_.getMessage())
+    verifySameDiagnostics(dpParseDiag, saxParseDiag)
 
     new DaffodilTDMLParseResult(actual, outputter)
   }
 
+  def doUnparseWithBothApis(dpInputter: TDMLInfosetInputter, saxInputStream: java.io.InputStream,
+    dpOutputStream: java.io.OutputStream): DaffodilTDMLUnparseResult = {
+
+    val dpOutputChannel = java.nio.channels.Channels.newChannel(dpOutputStream)
+    val saxOutputStream = new ByteArrayOutputStream
+    val saxOutputChannel = java.nio.channels.Channels.newChannel(saxOutputStream)
+    val errorHandler = new DaffodilTDMLSAXErrorHandler()
+    val inputContentHandler = new DaffodilInputContentHandler(dp, saxOutputChannel)
+    val xmlReader = SAXParserFactory.newInstance.newSAXParser.getXMLReader
+    xmlReader.setContentHandler(inputContentHandler)
+    xmlReader.setErrorHandler(errorHandler)
+    xmlReader.setFeature(XMLUtils.DAFFODIL_SAX_NAMESPACE_FEATURE, true)
+    xmlReader.setFeature(XMLUtils.DAFFODIL_SAX_NAMESPACE_PREFIX_FEATURE, true)
+
+    val actualDP = dp.unparse(dpInputter, dpOutputChannel).asInstanceOf[UnparseResult]
+    dpOutputChannel.close()
+    // kick off SAX Unparsing
+    try {
+      xmlReader.parse(new InputSource(saxInputStream))
+    } catch {
+      case e: DaffodilUnhandledSAXException =>
+        // In the case of an unexpected errors, catch and throw as TDMLException
+        throw TDMLException("Unexpected error during SAX Unparse:" + e, None)
+      case _: DaffodilUnparseErrorSAXException =>
+        // TODO not sure what else to do with this error, unparseResult and its diagnostics
+        // will be handled below
```

Comment:
Yeah, I think this is safe to ignore since you get the parse result from the ContentHandler.